File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    vector<string> ans;
    bool isValid(string &str){
        int cnt = 0;
        for(char ch : str){
            if(ch == '('){
                cnt++;
            }else{
                cnt--;
            }
            if(cnt < 0)
                return false;
        }
        return cnt == 0;
    }
    void solve(string& curr, int n){
        if(curr.length() == 2*n){
            if(isValid(curr)){
                ans.push_back(curr);
            }
            return;
        }

        // take open bracket
        curr.push_back('(');
        solve(curr, n);
        curr.pop_back();

        // take close bracket
        curr.push_back(')');
        solve(curr, n);
        curr.pop_back();
    }
    vector<string> generateParenthesis(int n) {
        string curr = "";
        solve(curr, n);

        return ans;
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public void solve(StringBuilder sb, int n, int open, int close, List<String> res){
        if(sb.length() == 2*n){
            res.add(sb.toString());
            return ;
        }
        if(open < n){
            sb.append('(');
            solve(sb, n, open+1, close, res);
            sb.deleteCharAt(sb.length()-1);
        }

        if(close < open){
            sb.append(')');
            solve(sb, n, open, close+1, res);
            sb.deleteCharAt(sb.length()-1);
        }
    }
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        StringBuilder sb = new StringBuilder();

        solve(sb, n, 0, 0, res);

        return res;
    }
}